pkg/cover/backend/elf.go

```diff
@@ -196,3 +176,43 @@ func elfReadModuleCoverPoints(target *targets.Target, module *Module, info *symb
 	}
 	return pcs, nil
 }
+
+func alignUp(addr, align uint64) uint64 {
+	if align == 0 {
+		return addr
+	}
+	return (addr + align - 1) & ^(align - 1)
+}
+
+// Calculate the offset of the module .text section in the kernel memory.
+// /proc/modules only contains the base address, which corresponds to the
+// beginning of the first code section, but that section does not have to
+// be .text (e.g. on Android it may be .plt).
+// The offset is calculated by summing up the aligned sizes of sections
+// that:
+//   - precede .text;
+//   - are not .init sections;
+//   - have the same section flags as .text.
+func elfGetModuleOffset(path string) uint64 {
+	file, err := elf.Open(path)
+	if err != nil {
+		return 0
+	}
+	defer file.Close()
+	ts := file.Section(".text")
+	if ts == nil {
+		return 0
+	}
+	off := uint64(0)
+	textFlags := ts.SectionHeader.Flags
+	for _, s := range file.Sections {
+		if s.Flags == textFlags && !strings.HasPrefix(s.SectionHeader.Name, ".init") {
+			off = alignUp(off, s.SectionHeader.Addralign)
+			if s == ts {
+				return off
+			}
+			off += s.SectionHeader.Size
+		}
+	}
+	return 0
```

Comment:
Given how subtle this is, how silent is the failure mode, and how hard it was to debug, can you think of a way to cross-test that we actually guessed the offset correctly and PCs will match in the end?
Otherwise we may mis-handle some corner cases even now (maybe some of the Android modules are still mis-handled), or things can break in future after syzkaller or kernel code changes.

Reply:
I had an idea to replace `elfGetModuleOffset` with a function that parses `/proc/kallsyms` calculates module offsets based on symbols listed there.
But it turns out we cannot do that for `syz-cover`, which has access to saved `/proc/modules`, but not `/proc/kallsyms`.

It also turns out that the contents of /proc/kallsyms are not visible at an arbitrary point and should be probably loaded together with /proc/modules during manager startup.

My plan is to load and parse /proc/kallsyms in `parseModules()` and store the name and offset inside .text of an arbitrary symbol for each module in `KernelModule`. Then we can probably use it to verify the symbol placement at some later point. This would however require carrying this name/offset pair around forever, and they aren't useful for anything else. WDYT?

Reply:
Hm, I am looking at the kallsyms contents, and it's starting to get trickier.

An important corner case is `pkvm-s2mpu-v9.ko`, which has the following sections:

```
  0 __ksymtab_gpl 00000054  0000000000000000  0000000000000000  00000040  2**2
                  CONTENTS, ALLOC, LOAD, RELOC, READONLY, DATA
  1 __kcrctab_gpl 0000001c  0000000000000000  0000000000000000  00000094  2**0
                  CONTENTS, ALLOC, LOAD, RELOC, READONLY, DATA
  2 .altinstructions 0000033c  0000000000000000  0000000000000000  000000b0  2**3
                  CONTENTS, ALLOC, LOAD, RELOC, READONLY, DATA
  3 __bug_table   00000090  0000000000000000  0000000000000000  000003f0  2**3
                  CONTENTS, ALLOC, LOAD, RELOC, DATA
  4 __jump_table  00000080  0000000000000000  0000000000000000  00000480  2**3
                  CONTENTS, ALLOC, LOAD, RELOC, DATA
  5 .plt          00000001  0000000000000000  0000000000000000  00000500  2**4
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
  6 .init.plt     00000001  0000000000000000  0000000000000000  00000501  2**0
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
  7 .text.ftrace_trampoline 00000001  0000000000000000  0000000000000000  00000502  2**0
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
  8 .hyp.text     00002000  0000000000000000  0000000000000000  00001000  2**12
                  CONTENTS, ALLOC, LOAD, RELOC, READONLY, CODE
  9 .hyp.bss      00001000  0000000000000000  0000000000000000  00003000  2**12
                  ALLOC
 10 .hyp.rodata   00001000  0000000000000000  0000000000000000  00003000  2**12
                  CONTENTS, ALLOC, LOAD, RELOC, READONLY, DATA
 11 .hyp.data     00001000  0000000000000000  0000000000000000  00004000  2**12
                  CONTENTS, ALLOC, LOAD, RELOC, DATA
 12 .hyp.reloc    00000048  0000000000000000  0000000000000000  00005000  2**2
                  CONTENTS, ALLOC, LOAD, RELOC, READONLY, DATA
 13 .text         00001aac  0000000000000000  0000000000000000  00005048  2**2
                  CONTENTS, ALLOC, LOAD, RELOC, READONLY, CODE
...
```

As you can see, here `.text` starts at 0x3000, because there's a bunch of small sections within the first 0x1000 followed by `.hyp.text` that is aligned on 0x1000 and has a size of 0x2000.
Now, from `/proc/modules` we can see that pkvm_s2mpu_v9 is allocated at 0xffffffc001847000:
```
69632 2 samsung_iommu_v9, Live 0xffffffc001847000
```

, but looking at `/proc/kallsyms`, we cannot really tell the difference between a symbol belonging to `.text` and e.g. `.hyp.text`:

```
# cat /proc/kallsyms  | grep pkvm_s2mpu_v9 | sort | head -n 100                                                                  
ffffffc001848000 t __kvm_nvhe_$x.1	[pkvm_s2mpu_v9]
ffffffc001848000 t __kvm_nvhe___hyp_section_.hyp.text	[pkvm_s2mpu_v9]
ffffffc001848000 t __kvm_nvhe_memset	[pkvm_s2mpu_v9]
ffffffc001848028 t __kvm_nvhe_memcpy	[pkvm_s2mpu_v9]
ffffffc001848050 t __kvm_nvhe_s2mpu_init	[pkvm_s2mpu_v9]
ffffffc0018481a8 t __kvm_nvhe_$x.4	[pkvm_s2mpu_v9]
ffffffc0018481ec t __kvm_nvhe_s2mpu_validate	[pkvm_s2mpu_v9]
ffffffc001848208 t __kvm_nvhe_s2mpu_validate_child	[pkvm_s2mpu_v9]
ffffffc00184822c t __kvm_nvhe_s2mpu_host_stage2_idmap_prepare	[pkvm_s2mpu_v9]
ffffffc00184828c t __kvm_nvhe_s2mpu_host_stage2_idmap_apply	[pkvm_s2mpu_v9]
ffffffc001848380 t __kvm_nvhe_s2mpu_host_stage2_idmap_complete	[pkvm_s2mpu_v9]
ffffffc001848424 t __kvm_nvhe_s2mpu_suspend	[pkvm_s2mpu_v9]
ffffffc0018484c0 t __kvm_nvhe_s2mpu_resume	[pkvm_s2mpu_v9]
ffffffc001848558 t __kvm_nvhe_s2mpu_host_dabt_handler	[pkvm_s2mpu_v9]
ffffffc001848624 t __kvm_nvhe_sysmmu_sync_validate	[pkvm_s2mpu_v9]
ffffffc001848668 t __kvm_nvhe_s2mpu_hyp_init	[pkvm_s2mpu_v9]
ffffffc0018486d0 t __kvm_nvhe___invalidation_barrier_slow	[pkvm_s2mpu_v9]
ffffffc00184881c t __kvm_nvhe___initialize_ctx	[pkvm_s2mpu_v9]
ffffffc0018489cc t __kvm_nvhe___all_invalidation	[pkvm_s2mpu_v9]
ffffffc001848acc t __kvm_nvhe_$x.6	[pkvm_s2mpu_v9]
ffffffc001848ad0 t __kvm_nvhe_$x.8	[pkvm_s2mpu_v9]
ffffffc001848ad4 t __kvm_nvhe_$x.10	[pkvm_s2mpu_v9]
ffffffc001848ad8 t __kvm_nvhe_$x.12	[pkvm_s2mpu_v9]
ffffffc001848adc t __kvm_nvhe_$x.14	[pkvm_s2mpu_v9]
ffffffc001848ae0 t __kvm_nvhe_$x.16	[pkvm_s2mpu_v9]
ffffffc001848ae4 t __kvm_nvhe_$x.18	[pkvm_s2mpu_v9]
ffffffc001848ae8 t __kvm_nvhe_$x.20	[pkvm_s2mpu_v9]
ffffffc001848aec t __kvm_nvhe_$x.22	[pkvm_s2mpu_v9]
ffffffc001848af0 t __kvm_nvhe_$x.24	[pkvm_s2mpu_v9]
ffffffc001848af4 t __kvm_nvhe_$x.26	[pkvm_s2mpu_v9]
ffffffc001848af8 t __kvm_nvhe_$x.28	[pkvm_s2mpu_v9]
ffffffc001848afc t __kvm_nvhe_$x.30	[pkvm_s2mpu_v9]
ffffffc001848b00 t __kvm_nvhe_$x.32	[pkvm_s2mpu_v9]
ffffffc001848b04 t __kvm_nvhe_$x.34	[pkvm_s2mpu_v9]
ffffffc001848b08 t __kvm_nvhe_$x.36	[pkvm_s2mpu_v9]
ffffffc001848b0c t __kvm_nvhe_$x.38	[pkvm_s2mpu_v9]
ffffffc001848b10 t __kvm_nvhe_$x.40	[pkvm_s2mpu_v9]
ffffffc001848b14 t __kvm_nvhe_$x.42	[pkvm_s2mpu_v9]
ffffffc001848b18 t __kvm_nvhe_$x.44	[pkvm_s2mpu_v9]
ffffffc001848b1c t __kvm_nvhe_$x.46	[pkvm_s2mpu_v9]
ffffffc001848b20 t __kvm_nvhe_$x.48	[pkvm_s2mpu_v9]
ffffffc001848b24 t __kvm_nvhe_$x.50	[pkvm_s2mpu_v9]
ffffffc001848b28 t __kvm_nvhe_$x.52	[pkvm_s2mpu_v9]
ffffffc001848b2c t __kvm_nvhe_$x.54	[pkvm_s2mpu_v9]
ffffffc001848b30 t __kvm_nvhe_$x.56	[pkvm_s2mpu_v9]
ffffffc001848b34 t __kvm_nvhe_$x.58	[pkvm_s2mpu_v9]
ffffffc001848b38 t __kvm_nvhe_$x.60	[pkvm_s2mpu_v9]
ffffffc001848b3c t __kvm_nvhe_$x.62	[pkvm_s2mpu_v9]
ffffffc001848b40 t __kvm_nvhe_$x.64	[pkvm_s2mpu_v9]
ffffffc001848b44 t __kvm_nvhe_$x.66	[pkvm_s2mpu_v9]
ffffffc001848b48 t __kvm_nvhe_$x.1	[pkvm_s2mpu_v9]
ffffffc001848b48 t __kvm_nvhe_s2mpu_get_mpt_ops	[pkvm_s2mpu_v9]
ffffffc001848b68 t __kvm_nvhe_smpt_size	[pkvm_s2mpu_v9]
ffffffc001848b78 t __kvm_nvhe_init_with_prot	[pkvm_s2mpu_v9]
ffffffc001848cc8 t __kvm_nvhe_init_with_mpt	[pkvm_s2mpu_v9]
ffffffc001848dc8 t __kvm_nvhe_apply_range	[pkvm_s2mpu_v9]
ffffffc001848e78 t __kvm_nvhe_prepare_range	[pkvm_s2mpu_v9]
ffffffc0018490c0 t __kvm_nvhe_pte_from_addr_smpt	[pkvm_s2mpu_v9]
ffffffc0018490e4 t __kvm_nvhe___set_smpt_range	[pkvm_s2mpu_v9]
ffffffc00184a000 t $x.1	[pkvm_s2mpu_v9]
ffffffc00184a000 t __pkvm_s2mpu_of_link	[pkvm_s2mpu_v9]
ffffffc00184a1e0 t __pkvm_s2mpu_of_parse	[pkvm_s2mpu_v9]
ffffffc00184a254 t __pkvm_s2mpu_suspend	[pkvm_s2mpu_v9]
ffffffc00184a2d0 t __pkvm_s2mpu_resume	[pkvm_s2mpu_v9]
ffffffc00184a340 t s2mpu_probe	[pkvm_s2mpu_v9]
ffffffc00184a600 t $x.14	[pkvm_s2mpu_v9]
```

So in order to verify the module placement we'd need to:
1. Store **all** symbols corresponding to `pkvm_s2mpu_v9` at boot time by parsing `/proc/kallsyms` in `parseModules()`. We cannot do this later, because `/proc/kallsyms` is not guaranteed to return the same (or even non-zero) addresses when we request the coverage.
2. In `elfGetModuleOffset()`, pick a symbol from the module's `.text` section and make sure one of the symbols with the same name has the expected address. We cannot pick a symbol in `parseModules()`, because discovery of modules happens later, when the user requests the coverage.

This is doable, but conceptually I am a bit worried about carrying around the parsed kallsyms data around.

Reply:
> but looking at /proc/kallsyms, we cannot really tell the difference between a symbol belonging to .text and e.g. .hyp.text:

Do we need to? If both sections are loaded as executable code, do we care about the difference?

Reply:
As far as I understand, we do.
E.g. `elfReadSymbols()` in `pkg/cover/backend/elf.go` reads the symbols from the `.text` section, not any other `.*.text` sections (which makes sense, because `.hyp.text` does not contain any sancov hooks). Therefore `module.Addr` must point to the beginning of `.text`, and it's important that we pick a symbol belonging to `.text` to verify that.

Reply:
> E.g. elfReadSymbols() in pkg/cover/backend/elf.go reads the symbols from the .text section, not any other .*.text sections (which makes sense, because .hyp.text does not contain any sancov hooks).

Other code sections are loaded and can contain coverage callbacks, no?
This may be a bug that we don't read other sections.

> Therefore module.Addr must point to the beginning of .text

I think it should point to beginning of the code segment (all loaded code sections).

Reply:
> Other code sections are loaded and can contain coverage callbacks, no?
> This may be a bug that we don't read other sections.

Out of 297 Android binaries I am looking at (1 vmlinux and 296 .ko files), here are the text sections present in them:

```
    244 .exit.text
      1 .head.text
    296 .hyp.text
    246 .init.text
    297 .text
    296 .text.ftrace_trampoline
```

I am not sure what is in `.head.text` (it mostly contains repeated byte patterns that objdump doesn't disassemble as valid text).
`.init.text` and `exit.text` are not interesting for us, because callbacks from there are not supposed to be called during fuzzing.
`.text.ftrace_trampoline` always has a size of 1 byte (i.e. no callbacks)
None of the `.hyp.text` sections contains any coverage callbacks.

So I am pretty confident we are good with just parsing `.text` for now.

> I think it should point to beginning of the code segment (all loaded code sections).
This is what is happening currently: it's a convention that all code sections are loaded in the same memory region, and lines in /proc/modules are pointing to that region. So without my patches `module.Addr` already points to that region. But then we'd have to calculate the offsets of every code section using the same code I propose in the patch to calculate the offset of `.text`. And we'd still want to cross-check it somehow?

Reply:
What I also noticed is that there's no single symbol name common for all modules.
E.g. there can be as few as 4 text symbols corresponding to a module:

```
$ cat /proc/kallsyms
...
ffffffc0018ad000 t $x.1	[vh_cgroup]
ffffffc0018ad034 t $x.1	[vh_cgroup]
ffffffc0018ad000 t init_module	[vh_cgroup]
ffffffc0018ad034 t redacted_name_foobar	[vh_cgroup]
```

```
$ nm vh_cgroup.ko | grep ' [Tt] '
0000000000000000 t $x.1
0000000000000034 t $x.1
0000000000000000 T init_module
0000000000000034 T redacted_name_foobar
```

Reply:
Any other ideas for testing?

What percent of coverage PCs we currently match against callbacks? Perhaps we could just ensure that we match all of them, or at least most of them, for each module? If we matched only 1% of 1000 PCs, there is definitely something broken somewhere.

Reply:
Hm, matching PCs is actually a good idea.

Am I getting it right we can do this in prepareFileMap() (https://github.com/google/syzkaller/blob/4f9530a3b62297342999c9097c77dde726522618/pkg/cover/report.go#L124) by calculating the numbers of matched PCs and comparing it to the total number of PCs in all programs?

Reply:
Turned out this code was matching not the coverage PCs, but rather frames, which was quite relaxed.
I added "pkg/cover: ensure that all PCs returned by kcov have matching callbacks" that builds a map of coverage PCs and compares them against the PCs returned by kcov.